File: data_structures/linked_list.py
class Node(object):
    def __init__(self, data):
        self.data = data
        self.next = None

class SinglyLinkedList(object):
    def __init__(self, data=None) -> None:
        self.head = Node(data) or None

    def add(self, data):
        new_node = Node(data)
        new_node.next = self.head
        self.head = new_node

    def search(self, data):
        if self.head == None:
            raise("LinkedList is empty")
        current = self.head
        while current:
            if current.data == data:
                return current
            current = current.next
        raise ValueError("Value not in Linked List")

    def size(self):
        index = 0
        current = self.head
        while current:
            index += 1
            current = current.next
        return index

    def remove(self, data):
        if self.head == None:
            raise ValueError("LinkedList is empty")
        current = self.head
        if current.data == data:
            self.head = current.next
            current = None
            return
        previous = None
        while current.next:
            previous = current
            current = current.next
            if current.data == data:
                previous.next = current.next
                

File: data_structures/linked_list.js
class Node {
  constructor(data) {
    this.data = data;
    this.next = null;
  }
}

class SinglyLinkedList {
  constructor(head = null) {
    if (head) {
      this.head = new Node(head);
    } else {
      this.head = null;
    }
  }

  add(item) {
    const new_node = new Node(item);
    if (this.head) {
      new_node.next = this.head;
      this.head = new_node;
    } else {
      this.head = new_node;
    }
  }

  print() {
    let current = this.head;
    let str = "";
    while (current) {
      str = str + " " + current.data;
      current = current.next;
    }
    return str.trim();
  }

  remove(data) {
    if (!this.head) throw new Error("linkedlist is empty");
    if (this.head.data == data) {
      this.head = this.head.next;
    } else {
      let prev;
      let current = this.head;
      while (current) {
        prev = current;
        current = current.next;
        if (current.data == data) {
          prev.next = current.next;
          current = null;
          return;
        }
      }
    }
  }
}

exports.SinglyLinkedList = SinglyLinkedList;
